pkg/cmd/provisioning/roles.go

```diff
@@ -0,0 +1,337 @@
+	credRequests := []*credreqv1.CredentialsRequest{}
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, file := range files {
+		f, err := os.Open(filepath.Join(dir, file.Name()))
+		if err != nil {
+			return nil, errors.Wrap(err, "Failed to open file")
+		}
+		defer f.Close()
+		decoder := yaml.NewYAMLOrJSONDecoder(f, 4096)
+		for {
+			cr := &credreqv1.CredentialsRequest{}
+			if err := decoder.Decode(cr); err != nil {
+				if err == io.EOF {
+					break
+				}
+				return nil, errors.Wrap(err, "Failed to decode to CredentialsRequest")
+			}
+			credRequests = append(credRequests, cr)
+		}
+
+	}
+
+	return credRequests, nil
+}
+
+func getIssuerURLFromIdentityProvider(awsClient aws.Client, idProviderARN string) (string, error) {
+	idProvider, err := awsClient.GetOpenIDConnectProvider(&iam.GetOpenIDConnectProviderInput{
+		OpenIDConnectProviderArn: awssdk.String(idProviderARN),
+	})
+
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get IAM Identity Provider")
+	}
+
+	return *idProvider.Url, nil
+}
+
+func iamRolesCmd(cmd *cobra.Command, args []string) {
+	cfg := &awssdk.Config{
+		Region: awssdk.String(CreateOpts.Region),
+	}
+
+	s, err := session.NewSession(cfg)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	awsClient := aws.NewClientFromSession(s)
+
+	err = createIAMRoles(awsClient, CreateOpts.IdentityProviderARN, CreateOpts.NamePrefix, CreateOpts.CredRequestDir, CreateOpts.TargetDir, CreateOpts.GenerateOnly)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+// StatementEntry is a simple type used to serialize to AWS' PolicyDocument format.
+type StatementEntry struct {
+	Effect   string
+	Action   []string
+	Resource string
+	// Must "omitempty" otherwise we send unacceptable JSON to the AWS API when no
+	// condition is defined.
+	Condition credreqv1.IAMPolicyCondition `json:",omitempty"`
+}
+
+// PolicyDocument is a simple type used to serialize to AWS' PolicyDocument format.
+type PolicyDocument struct {
+	Version   string
+	Statement []StatementEntry
+}
+
+func createRolePolicy(statements []credreqv1.StatementEntry) string {
+	policyDocument := PolicyDocument{
+		Version:   "2012-10-17",
+		Statement: []StatementEntry{},
+	}
+
+	for _, entry := range statements {
+		policyDocument.Statement = append(policyDocument.Statement,
+			StatementEntry{
+				Effect:    entry.Effect,
+				Action:    entry.Action,
+				Resource:  entry.Resource,
+				Condition: entry.PolicyCondition,
+			})
+	}
+
+	b, err := json.Marshal(&policyDocument)
+	if err != nil {
+		log.Fatalf("Failed to marshal the policy to JSON: %s", err)
+	}
+
+	return string(b)
+}
+
+// NewIAMRolesSetup provides the "create iam-roles" subcommand
+func NewIAMRolesSetup() *cobra.Command {
+	iamRolesSetupCmd := &cobra.Command{
+		Use: "iam-roles",
+		Run: iamRolesCmd,
+	}
+
+	iamRolesSetupCmd.PersistentFlags().StringVar(&CreateOpts.NamePrefix, "name-prefix", "", "Name prefix for all created AWS resources")
```

Comment:
With sts-preflight and this, I wasn't totally clear what this name prefix was meant to be. Do I just make it up? It feels like infraID but that's not known until later. Presumably this is just something the user can pick freeform? 

If so lets just note that in the description.